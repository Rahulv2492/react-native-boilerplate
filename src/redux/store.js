import createSagaMiddleware from 'redux-saga';
import reducers from './reducers';
import rootSaga from './sagas';
import { createStore, applyMiddleware } from 'redux';

const sagaMiddleware = createSagaMiddleware();

const store = createStore(reducers, applyMiddleware(sagaMiddleware));

sagaMiddleware.run(rootSaga);

export default store;
