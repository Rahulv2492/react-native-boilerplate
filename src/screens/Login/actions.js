import { USER_API } from '../../constants/ActionTypes';

export const userLogin = () => {
    return {
        type: USER_API.REQUESTING,
        state: {
            isAuthenticated: false,
            user: {
                username: '<username>',
                password: '<Password>',
            },
        },
    };
};

export const restoreToken = playload => {
    return {
        type: USER_API.RESTORE_TOKEN,
        state: { ...playload },
    };
};
