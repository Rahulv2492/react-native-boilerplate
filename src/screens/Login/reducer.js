import { USER_API } from './../../constants/ActionTypes';
const LoginReducer = (state = {}, action) => {
    switch (action.type) {
        case USER_API.RESTORE_TOKEN:
            return {
                ...state,
                ...action.state,
            };
        default:
            return state;
    }
};

export default LoginReducer;
