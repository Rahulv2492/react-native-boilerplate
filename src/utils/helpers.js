import AsyncStorage from '@react-native-community/async-storage';
import * as Promise from 'bluebird';

const catchLogThrow = err => {
    console.error(err);
    return Promise.reject(err);
};

const storeData = obj => {
    return Promise.map(Object.keys(obj), key =>
        AsyncStorage.setItem(key, obj[key]),
    ).catch(catchLogThrow);
};

const retrieveData = key => {
    return AsyncStorage.getItem(key).catch(catchLogThrow);
};

const createRequestTypes = base => {
    const requestStatus = ['REQUESTING', 'SUCCESS', 'FAILED', 'INVALIDATE'];

    return requestStatus.reduce((acc, type) => {
        acc[type] = `${type}_${base}`;
        return acc;
    }, {});
};

const serialize = obj => {
    return Object.keys(obj)
        .map(
            key => `${encodeURIComponent(key)}=${encodeURIComponent(obj[key])}`,
        )
        .join('&');
};
export { storeData, retrieveData, createRequestTypes, serialize };
