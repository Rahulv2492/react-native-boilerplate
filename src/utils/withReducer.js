const withReducer = (reducer_name, reducer = null) => (state = {}, action) => {
    switch (action.type) {
        case `REQUESTING_${reducer_name}`:
            return {
                ...state,
                ...action.state,
                isLoading: true,
                success: false,
                error: false,
            };
        case `SUCCESS_${reducer_name}`:
            return {
                ...state,
                ...action.state,
                isLoading: false,
                success: true,
                error: false,
            };
        case `FAILED_${reducer_name}`:
            return {
                ...state,
                ...action.state,
                isLoading: false,
                success: false,
                error: true,
            };
        case `INVALIDATE_${reducer_name}`:
            return {
                ...state,
                ...action.state,
                isLoading: false,
                success: false,
                error: false,
            };
        default:
            return reducer ? reducer(state, action) : state;
    }
};

export default withReducer;
