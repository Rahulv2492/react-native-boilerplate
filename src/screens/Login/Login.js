import React from 'react';
import _ from 'lodash';
import PropTypes from 'prop-types';
import { Text, Button } from 'react-native';
import { connect } from 'react-redux';
import { userLogin } from './actions';
import { Container } from './Login.style';

const Login = ({ userLogin }) => {
    return (
        <Container.Wrapper>
            <Text>Login Page</Text>
            <Button onPress={() => userLogin()} title="Login" />
        </Container.Wrapper>
    );
};

Login.propTypes = {
    userLogin: PropTypes.func,
};

const mapDispatchToProps = dispatch => ({
    userLogin: () => dispatch(userLogin()),
});

export default connect(
    null,
    mapDispatchToProps,
)(Login);
