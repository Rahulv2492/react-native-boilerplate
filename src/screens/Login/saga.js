import { takeEvery, call, put, all } from 'redux-saga/effects';
import { USER_API } from '../../constants/ActionTypes';
import { AUTHRIZATION_TOKEN_KEY } from '../../constants';
import request from '../../services/RequestService';
import { storeData } from '../../utils/helpers';
import _ from 'lodash';

function* signIn(action) {
    const userBody = _.get(action, 'state.user');
    try {
        // const {
        //     data: { user, token },
        // } = yield call(request, {
        //     url: `auth/login`,
        //     method: 'POST',
        //     data: userBody,
        // });
        const token = 'demo token';
        const user = { name: 'demo' };
        if (token && token.length) {
            yield storeData({
                [AUTHRIZATION_TOKEN_KEY]: `Bearer ${token}`,
            });
        }
        yield put({
            type: USER_API.SUCCESS,
            state: {
                isAuthenticated: true,
                user,
            },
        });
    } catch (error) {
        yield put({
            type: USER_API.FAILED,
            state: {
                isAuthenticated: false,
                user: {},
            },
        });
    }
}

export default function* rootSaga() {
    yield all([yield takeEvery(USER_API.REQUESTING, signIn)]);
}
