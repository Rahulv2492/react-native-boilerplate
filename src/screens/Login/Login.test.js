import React from 'react';

import Login from './Login';
import store from './../../redux/store';
import { render } from 'react-native-testing-library';
import { createRequestTypes } from './../../utils/helpers';

describe('Login', () => {
    it('should match to snapshot', () => {
        const tree = render(<Login store={store} />).toJSON();
        console.log('tree', tree);
        expect(tree).toMatchSnapshot();
    });

    it('should match to request types', () => {
        expect(createRequestTypes('DEMO')).toStrictEqual({
            REQUESTING: 'REQUESTING_DEMO',
            SUCCESS: 'SUCCESS_DEMO',
            FAILED: 'FAILED_DEMO',
            INVALIDATE: 'INVALIDATE_DEMO',
        });
    });
});
