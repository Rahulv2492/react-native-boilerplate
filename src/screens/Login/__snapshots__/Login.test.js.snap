// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login should match to snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
        "justifyContent": "center",
      },
    ]
  }
>
  <Text>
    Login Page
  </Text>
  <View
    accessibilityRole="button"
    accessibilityState={Object {}}
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Login
      </Text>
    </View>
  </View>
</View>
`;
